pkg/reconciler/sequence/sequence.go

```diff
@@ -333,3 +341,149 @@ func (r *Reconciler) removeUnwantedSubscriptions(ctx context.Context, seq *v1.Se
 
 	return nil
 }
+
+func (r *Reconciler) reconcileEventPolicies(ctx context.Context, s *v1.Sequence, channels []*eventingduckv1.Channelable, subs []*messagingv1.Subscription, featureFlags feature.Flags) error {
+	if !featureFlags.IsOIDCAuthentication() {
+		return r.cleanupAllEventPolicies(ctx, s)
+	}
+
+	existingPolicies, err := r.listEventPoliciesForSequence(s)
+	if err != nil {
+		return fmt.Errorf("failed to list existing EventPolicies: %w", err)
+	}
+
+	// Prepare maps for efficient lookups, updates, and deletions of policies
+	existingPolicyMap := make(map[string]*eventingv1alpha1.EventPolicy)
+	for _, policy := range existingPolicies {
+		existingPolicyMap[policy.Name] = policy
+	}
+
+	// Prepare lists for different actions so that policies can be categorized
+	var policiesToUpdate, policiesToCreate []*eventingv1alpha1.EventPolicy
+	policiesToDelete := make([]*eventingv1alpha1.EventPolicy, 0, len(existingPolicies))
+
+	// Handle intermediate channel policies (skip the first channel as it's the input channel!)
+	for i := 1; i < len(channels); i++ {
+		expectedPolicy := resources.MakeEventPolicyForSequenceChannel(s, channels[i], subs[i-1])
+		existingPolicy, exists := existingPolicyMap[expectedPolicy.Name]
+
+		if exists {
+			if !equality.Semantic.DeepDerivative(expectedPolicy, existingPolicy) {
+				expectedPolicy.SetResourceVersion(existingPolicy.ResourceVersion)
+				policiesToUpdate = append(policiesToUpdate, expectedPolicy)
+			}
+			delete(existingPolicyMap, expectedPolicy.Name)
+		} else {
+			policiesToCreate = append(policiesToCreate, expectedPolicy)
+		}
+	}
+
+	// Handle input channel policies
+	inputPolicies, err := r.prepareInputChannelEventPolicy(s, channels[0])
+	if err != nil {
+		return fmt.Errorf("failed to prepare input channel EventPolicies: %w", err)
+	}
+	for _, inputPolicy := range inputPolicies {
+		existingInputPolicy, exists := existingPolicyMap[inputPolicy.Name]
+		if exists {
+			if !reflect.DeepEqual(inputPolicy, existingInputPolicy) {
```

Comment:
Why was this changed from `equality.Semantic.DeepDerivative`? `DeepDerivative(inputPolicy, existingInputPolicy)` ignores empty fields in `inputPolicy`, which can be useful, as `existingInputPolicy` probably has some updated `.status` or annotations added by k8s (e.g. last-applied-configuration annotation).

Reply:
the deepDerivative failed to compare the eventpolicy's From.Spec

See the [latest test case](https://github.com/knative/eventing/pull/8106/commits/6042cbbd95fb7018ded81ee7edf70000869d934a) I added. @creydr

Reply:
ok, but the existingInputPolicy could have some additional annotations (e.g. added by some 3rd party), which we don't care about and do not require an update of the policy. That is the benefit of `equality.Semantic.DeepDerivative(a1, a2)`, which allows to ignore unset fields in a1:

> DeepDerivative is similar to DeepEqual except that unset fields in a1 are ignored (not compared). This allows us to focus on the fields that matter to the semantic comparison.

Your unit test is passing, because you only have some updates on the spec. But when your policy would have a valid .spec, but some .status or some unimportant annotations which would actually not require an update, your code would IIUC still trigger an update

Reply:
Ohh mb I didn't think about from that perspective! Thanks for the insight Christoph! @creydr 

I will make the changes accordingly.